### PR TITLE
Added optional password parameter to signUp/signIn methods

### PR DESCRIPTION
**Pull request by borgesmb**
## Proposed changes

- Removed SignUpUsingPassword method from public interface
- Removed SignInUsingPassword method from public interface
- Unified SignUpPasswordStartError and SignUpStartError
- Unified SignUpPasswordStartError and SignUpStartError
- Unified SignInStartDelegate and SignInPasswordStartDelegate
- Unified SignUpStartDelegate and SignUpPasswordStartDelegate
- Unified SignInPasswordStartResult and SignInStartResult
- Unified SignUpPasswordStartResult and SignUpStartResult
- Unified SignUpPasswordStartDelegateDispatcher and SignUpStartDelegateDispatcher
- Unified SignInPasswordStartDelegateDispatcher and SignInStartDelegateDispatcher

## Type of change

- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [X] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [X] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: MSAL/src/native_auth/controllers/sign_in/MSALNativeAuthSignInControlling.swift [hunk 1]
```diff
@@ -25,17 +25,17 @@
...
+
+    typealias SignInControllerResponse = MSALNativeAuthControllerTelemetryWrapper<SignInStartResult>
     typealias SignInAfterPreviousFlowControllerResponse =
         MSALNativeAuthControllerTelemetryWrapper<Result<MSALNativeAuthUserAccountResult, MSALNativeAuthError>>
     typealias SignInSubmitCodeControllerResponse = MSALNativeAuthControllerTelemetryWrapper<SignInVerifyCodeResult>
     typealias SignInSubmitPasswordControllerResponse = MSALNativeAuthControllerTelemetryWrapper<SignInPasswordRequiredResult>
     typealias SignInResendCodeControllerResponse = MSALNativeAuthControllerTelemetryWrapper<SignInResendCodeResult>
 
-    func signIn(params: MSALNativeAuthSignInWithPasswordParameters) async -> SignInPasswordControllerResponse
+    func signIn(params: MSALNativeAuthSignInWithPasswordParameters) async -> SignInControllerResponse
```
Comment (nilo-ms) [inline MSAL/src/native_auth/controllers/sign_in/MSALNativeAuthSignInControlling.swift:36]: do you think would be possible to use only one param class in this case? So, we can have only one signIn method

File: MSAL/src/native_auth/network/errors/sign_up/MSALNativeAuthSignUpChallengeResponseError.swift [hunk 2]
```diff
@@ -42,7 +42,7 @@ struct MSALNativeAuthSignUpChallengeResponseError: MSALNativeAuthResponseError {
 
 extension MSALNativeAuthSignUpChallengeResponseError {
 
-    func toSignUpPasswordStartPublicError() -> SignUpPasswordStartError {
+    func toSignUpPasswordStartPublicError() -> SignUpStartError {
```
Comment (nilo-ms) [inline MSAL/src/native_auth/network/errors/sign_up/MSALNativeAuthSignUpChallengeResponseError.swift:45]: this method seems a duplicate of `toSignUpStartPublicError()` method

File: MSAL/src/native_auth/network/errors/sign_up/MSALNativeAuthSignUpStartResponseError.swift [hunk 3]
```diff
@@ -49,7 +49,7 @@ struct MSALNativeAuthSignUpStartResponseError: MSALNativeAuthResponseError {
 
 extension MSALNativeAuthSignUpStartResponseError {
 
-    func toSignUpStartPasswordPublicError() -> SignUpPasswordStartError {
+    func toSignUpStartPasswordPublicError() -> SignUpStartError {
```
Comment (nilo-ms) [inline MSAL/src/native_auth/network/errors/sign_up/MSALNativeAuthSignUpStartResponseError.swift:52]: Do you think this method and `toSignUpStartPublicError()` can be merged? Given that we're merging both signUp methods, we don't have any more a signUp password public error.

File: MSAL/src/native_auth/network/responses/validator/sign_in/validated_response/MSALNativeAuthSignInChallengeValidatedResponse.swift [hunk 4]
```diff
@@ -59,7 +59,7 @@ enum MSALNativeAuthSignInChallengeValidatedErrorType: Error {
         }
     }
 
-    func convertToSignInPasswordStartError() -> SignInPasswordStartError {
+    func convertToSignInPasswordStartError() -> SignInStartError {
```
Comment (nilo-ms) [inline MSAL/src/native_auth/network/responses/validator/sign_in/validated_response/MSALNativeAuthSignInChallengeValidatedResponse.swift:62]: This method name seems a little bit unclear to me now, we're saying that this method converts to SignInPassword but actually it converts it to SignInStartError. It's seems a little bit unconsistent to me

File: MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication+Internal.swift [hunk 5]
```diff
@@ -26,17 +26,20 @@ import Foundation
...
-    ) async -> MSALNativeAuthSignUpControlling.SignUpStartPasswordControllerResponse {
+    ) async -> MSALNativeAuthSignUpControlling.SignUpStartControllerResponse {
         guard inputValidator.isInputValid(username) else {
-            return .init(.error(SignUpPasswordStartError(type: .invalidUsername)))
+            return .init(.error(SignUpStartError(type: .invalidUsername)))
         }
-        guard inputValidator.isInputValid(password) else {
-            return .init(.error(SignUpPasswordStartError(type: .invalidPassword)))
+
+        if let password = password {
```
Comment (nilo-ms) [inline MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication+Internal.swift:39]: in terms of stile here, I'd put the if let and password check in the same if block, something like:
```suggestion
        if let password = password, !inputValidator.isInputValid(password) {
            return .init(.error(SignUpStartError(type: .invalidPassword)))
        }
```

File: MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication+Internal.swift [hunk 6]
```diff
@@ -49,72 +52,48 @@ extension MSALNativeAuthPublicClientApplication {
...
-        let params = MSALNativeAuthSignInWithPasswordParameters(
-            username: username,
-            password: password,
-            context: MSALNativeAuthRequestContext(correlationId: correlationId),
-            scopes: scopes
-        )
-
-        return await controller.signIn(params: params)
-    }
+        if let password = password {
```
Comment (nilo-ms) [inline MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication+Internal.swift:80]: Do we still need this distinction? What if we have only one signIn controller method?

File: MSAL/src/native_auth/public/state_machine/delegate/SignInDelegates.swift [hunk 7]
```diff
@@ -70,6 +46,11 @@ public protocol SignInStartDelegate {
     /// - Note: If a flow requires this optional method and it is not implemented, then ``onSignInStartError(error:)`` will be called.
     /// - Parameter newState: An object representing the new state of the flow with follow on methods.
     @MainActor @objc optional func onSignInPasswordRequired(newState: SignInPasswordRequiredState)
+    
+    /// Notifies the delegate that the sign in operation completed successfully.
+    /// - Parameter result: An object representing the signed in user account.
+    /// - Note: If a flow requires this optional method and it is not implemented, then ``onSignInPasswordStartError(error:)`` will be called.
```
Comment (nilo-ms) [inline MSAL/src/native_auth/public/state_machine/delegate/SignInDelegates.swift:52]: This comment still mention `onSignInPasswordStartError` method

File: MSAL/src/native_auth/controllers/sign_up/MSALNativeAuthSignUpControlling.swift [hunk 8]
```diff
@@ -26,16 +26,15 @@
...
     typealias SignUpResendCodeControllerResponse = MSALNativeAuthControllerTelemetryWrapper<SignUpResendCodeResult>
     typealias SignUpSubmitCodeControllerResponse = MSALNativeAuthControllerTelemetryWrapper<SignUpVerifyCodeResult>
     typealias SignUpSubmitPasswordControllerResponse = MSALNativeAuthControllerTelemetryWrapper<SignUpPasswordRequiredResult>
     typealias SignUpSubmitAttributesControllerResponse = MSALNativeAuthControllerTelemetryWrapper<SignUpAttributesRequiredResult>
 
-    func signUpStartPassword(parameters: MSALNativeAuthSignUpStartRequestProviderParameters) async -> SignUpStartPasswordControllerResponse
+    func signUpStartPassword(parameters: MSALNativeAuthSignUpStartRequestProviderParameters) async -> SignUpStartControllerResponse
 
-    func signUpStartCode(parameters: MSALNativeAuthSignUpStartRequestProviderParameters) async -> SignUpStartCodeControllerResponse
+    func signUpStartCode(parameters: MSALNativeAuthSignUpStartRequestProviderParameters) async -> SignUpStartControllerResponse
```
Comment (diegojerezba) [inline MSAL/src/native_auth/controllers/sign_up/MSALNativeAuthSignUpControlling.swift:37]: I leave the same [comment that I left in the abandoned PR](https://github.com/AzureAD/microsoft-authentication-library-for-objc/pull/1935#discussion_r1422913042) to keep track of it.

For clarification, I wasn't asking for a full refactor of the controller, just its interface. I see that @nilo-ms has [commented](https://github.com/AzureAD/microsoft-authentication-library-for-objc/pull/1942#discussion_r1427780228) something similar for `SignInController`.

File: MSAL/src/native_auth/public/state_machine/error/PasswordRequiredError.swift [hunk 9]
```diff
@@ -39,7 +39,7 @@ public class PasswordRequiredError: MSALNativeAuthError {
         super.init(message: message)
     }
 
-    init(signInPasswordError: SignInPasswordStartError) {
+    init(signInPasswordError: SignInStartError) {
```
Comment (diegojerezba) [inline MSAL/src/native_auth/public/state_machine/error/PasswordRequiredError.swift:42]: nit: the variable `signInPasswordError` can be renamed to match its new type

File: MSAL/test/unit/native_auth/public/error/SignUpStartErrorTests.swift [hunk 10]
```diff
@@ -30,7 +30,7 @@ final class SignUpStartErrorTests: XCTestCase {
     private var sut: SignUpStartError!
 
     func test_totalCases() {
-        XCTAssertEqual(SignUpStartError.ErrorType.allCases.count, 4)
+        XCTAssertEqual(SignUpStartError.ErrorType.allCases.count, 5)
```
Comment (diegojerezba) [inline MSAL/test/unit/native_auth/public/error/SignUpStartErrorTests.swift:33]: This test was created to remind us that if we add new errors, we should add a test for those errors 🙂.

The test `test_defaultErrorDescription` should be updated with the new error as well.

File: MSAL/test/unit/native_auth/public/error/SignUpPasswordStartErrorTests.swift
Comment (diegojerezba) [inline MSAL/test/unit/native_auth/public/error/SignUpPasswordStartErrorTests.swift:1]: These tests are testing the same class as `SignUpStartErrorTests`

File: MSAL/test/unit/native_auth/public/error/SignInPasswordStartErrorTests.swift
Comment (diegojerezba) [inline MSAL/test/unit/native_auth/public/error/SignInPasswordStartErrorTests.swift:1]: These tests are testing the same class as `SignInStartErrorTests`

File: MSAL/test/unit/native_auth/public/error/SignUpPasswordStartErrorTests.swift
Comment (nilo-ms) [reply MSAL/test/unit/native_auth/public/error/SignUpPasswordStartErrorTests.swift:1]: Thanks Diego, updated ;)

File: MSAL/test/unit/native_auth/public/error/SignInPasswordStartErrorTests.swift
Comment (nilo-ms) [reply MSAL/test/unit/native_auth/public/error/SignInPasswordStartErrorTests.swift:1]: Thanks Diego, updated ;)

File: MSAL/test/unit/native_auth/public/error/SignUpStartErrorTests.swift (hunk 10 above)
Comment (nilo-ms) [reply MSAL/test/unit/native_auth/public/error/SignUpStartErrorTests.swift:33]: Thanks Diego, updated ;)

File: MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication+Internal.swift (hunk 5 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication+Internal.swift]: Updated

File: MSAL/src/native_auth/public/state_machine/delegate/SignInDelegates.swift (hunk 7 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/public/state_machine/delegate/SignInDelegates.swift]: Updated

File: MSAL/src/native_auth/public/state_machine/error/PasswordRequiredError.swift (hunk 9 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/public/state_machine/error/PasswordRequiredError.swift]: Thanks Diego, updated ;)

File: MSAL/src/native_auth/network/errors/sign_up/MSALNativeAuthSignUpChallengeResponseError.swift (hunk 2 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/network/errors/sign_up/MSALNativeAuthSignUpChallengeResponseError.swift]: Duplicated method removed

File: MSAL/src/native_auth/controllers/sign_in/MSALNativeAuthSignInControlling.swift (hunk 1 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/controllers/sign_in/MSALNativeAuthSignInControlling.swift]: Updated, signUp, signIn code and password methods unified

File: MSAL/src/native_auth/network/errors/sign_up/MSALNativeAuthSignUpStartResponseError.swift (hunk 3 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/network/errors/sign_up/MSALNativeAuthSignUpStartResponseError.swift]: Methods merged

File: MSAL/src/native_auth/network/responses/validator/sign_in/validated_response/MSALNativeAuthSignInChallengeValidatedResponse.swift (hunk 4 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/network/responses/validator/sign_in/validated_response/MSALNativeAuthSignInChallengeValidatedResponse.swift]: These two methods were almost identical, removed one

File: MSAL/src/native_auth/controllers/sign_up/MSALNativeAuthSignUpControlling.swift (hunk 8 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/controllers/sign_up/MSALNativeAuthSignUpControlling.swift]: Just completed a full refactoring :)

**Review by diegojerezba — APPROVED** [review]

File: MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication.swift [hunk 11]
```diff
@@ -116,57 +116,24 @@ public final class MSALNativeAuthPublicClientApplication: MSALPublicClientApplic
     /// Sign up a user with a given username and password.
```
Comment (spetrescu84) [inline MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication.swift:116]: Is there a better way to describe this? As in "Sign up a user with a given username using OTP or Password"? Or just "Sign up a user with a given username"|

File: MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication.swift [hunk 12]
```diff
@@ -187,59 +154,22 @@ public final class MSALNativeAuthPublicClientApplication: MSALPublicClientApplic
 
     /// Sign in a user with a given username and password.
```
Comment (spetrescu84) [inline MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication.swift:155]: Same here, it's not necessarily a password

**Review by spetrescu84 — APPROVED** [review]

File: MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication.swift (hunk 11 above)
Comment (nilo-ms) [reply MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication.swift:116]: I'd like to mention also the word "password" in this comment.  About your suggestion:
"Sign up a user with a given username using OTP or Password", with that "or" it seems that the OTP code can be substituted by the password, but an OTP is always required during sign up right now.
"Sign up a user with a given username", we should mention that a user can specifies a password.

I'd prefer to keep the description as it is, and specifies in the parameter list that the password is optional

Comment (spetrescu84) [reply MSAL/src/native_auth/public/MSALNativeAuthPublicClientApplication.swift:116]: Ok, we can keep as is

**Pull request merged**